source/agora/consensus/Quorum.d

```diff
@@ -13,181 +13,476 @@
+import scpd.types.Stellar_types : uint256, NodeID;
+import scpd.types.Utils;
+import scpd.types.XDRBase;
+import scpd.quorum.QuorumTracker;
+
+import ocean.core.Test;
 
 import std.algorithm;
 import std.array;
+import std.conv;
 import std.exception;
 import std.format;
 import std.math;
+import std.random;
 import std.range;
 import std.string;
+import std.typecons;
+
+version (unittest)
+{
+    import agora.utils.Test;
+    import std.stdio;
+}
+
+/// Maximum number of nodes to include in a quorum.
+private enum MAX_NODES_IN_QUORUM = 7;
+
+/// Threshold to use for the quorum configuration (e.g. 0.80 = 80%)
+private enum double THRESHOLD = 0.80;
 
 /*******************************************************************************
 
-    Build the quorum configuration for the given public key and the staked
-    enrollments.
+    Build the quorum configuration for the given public key and the registered
+    enrollment keys.
 
     Params:
-        node_key = the key of the node for which to generate the quorum
-        keys = the array of registered enrollment UTXO keys
+        key = the key of the node for which to generate the quorum
+        utxo_keys = the list of UTXO keys of all the active enrollments
         finder = UTXO finder delegate
 
-    Returns:
-        the generated quorum config
-
 *******************************************************************************/
 
-public QuorumConfig buildQuorumConfig ( const ref PublicKey node_key,
-    in Hash[] keys, in UTXOFinder finder )
+public QuorumConfig buildQuorumConfig ( const ref PublicKey key,
+    in Hash[] utxo_keys, UTXOFinder finder )
 {
-    Set!PublicKey all_nodes = getPublicKeys(keys, finder);
+    // special-case: only 1 validator is active
+    if (utxo_keys.length == 1)
+        return QuorumConfig(1, [key]);
+
+    NodeStake[] stakes = buildStakesDescending(key, utxo_keys, finder);
 
     QuorumConfig quorum;
-    all_nodes.each!(node => quorum.nodes ~= node);
-    quorum.nodes.sort();
-    quorum.threshold = quorum.nodes.length;
+    quorum.nodes ~= key;  // add ourself first
+
+    // for filtering duplicates from dice()
+    auto added = BitField!uint(stakes.length);
+    auto rnd_gen = getGenerator(key);
+    auto stake_amounts = stakes.map!(stake => stake.amount.integral);
+
+    while (quorum.nodes.length < MAX_NODES_IN_QUORUM)
+    {
+        auto idx = dice(rnd_gen, stake_amounts);
+        if (added[idx])  // skip duplicate
+            continue;
 
+        quorum.nodes ~= stakes[idx].key;
+        added[idx] = true;  // mark used
+
+        if (quorum.nodes.length >= stakes.length + 1)  // our stake not included
+            break;  // ran out of nodes
+    }
+
+    quorum.nodes.sort;
+    quorum.threshold = max(1, cast(size_t)ceil(THRESHOLD * quorum.nodes.length));
     return quorum;
 }
 
+version (unittest)
+private const(PublicKey)[] getKeys (size_t count)
+{
+    return WK.Keys.byRange.take(count).map!(kp => kp.address).array;
+}
+
+/// 1 node
```

Comment:
Only one of these unittests should be ddoc-documented, the rest doesn't have to be.